File: apps/frontend/src/app/foundation/locations/location.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { ApiClient, ApiPage } from '@foundation/networking';
import {
  GeocodeSuggestion,
  LocationSearchFilters,
  LocationSearchResult,
  ParkingLocation,
  ParkingResource,
} from './location.models';

@Injectable({ providedIn: 'root' })
export class LocationService {
  private readonly api = inject(ApiClient);

  search(filters: LocationSearchFilters): Observable<ApiPage<LocationSearchResult>> {
    return this.api.get<ApiPage<LocationSearchResult>>('/locations/search', {
      params: filters,
    });
  }

  geocode(query: string): Observable<GeocodeSuggestion[]> {
    return this.api.get<GeocodeSuggestion[]>('/locations/geocode', {
      params: {
        query,
      },
    });
  }

  getLocation(locationId: string): Observable<ParkingLocation> {
    return this.api.get<ParkingLocation>(`/locations/${encodeURIComponent(locationId)}`);
  }

  listResources(locationId: string): Observable<ParkingResource[]> {
    return this.api.get<ParkingResource[]>(
      `/locations/${encodeURIComponent(locationId)}/resources`,
    );
  }
}

File: apps/ios/SpotLink/Sources/SpotLink/Features/Locations/LocationService.swift
import Foundation

// MARK: - Location Service

public final class LocationService: Sendable {
    private let apiClient: APIClientProtocol

    public init(apiClient: APIClientProtocol) {
        self.apiClient = apiClient
    }

    public func search(_ filters: LocationSearchFilters) async throws -> APIPage<LocationSearchResult> {
        try await apiClient.get("/locations/search", query: filters.queryParameters)
    }

    public func geocode(_ query: String) async throws -> [GeocodeSuggestion] {
        try await apiClient.get("/locations/geocode", query: ["query": query])
    }

    public func getLocation(_ locationId: String) async throws -> ParkingLocation {
        try await apiClient.get("/locations/\(locationId)", query: nil)
    }

    public func listResources(locationId: String) async throws -> [ParkingResource] {
        try await apiClient.get("/locations/\(locationId)/resources", query: nil)
    }
}
